core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -145,39 +152,31 @@ public BlockInStream getInStream(long blockId, Protocol.OpenUfsBlockOptions open
           Preconditions.checkNotNull(options.getUfsReadLocationPolicy(),
               PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
       address = blockLocationPolicy
-          .getWorker(GetWorkerOptions.defaults().setBlockWorkerInfos(getWorkerInfoList())
+          .getWorker(GetWorkerOptions.defaults().setBlockWorkerInfos(getAvailableWorkers())
               .setBlockId(blockId).setBlockSize(blockInfo.getLength()));
-      if (address == null) {
-        throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
-      }
     } else {
       // TODO(calvin): Get location via a policy.
-      // Although blockInfo.locations are sorted by tier, we prefer reading from the local worker.
-      // But when there is no local worker or there are no local blocks, we prefer the first
-      // location in blockInfo.locations that is nearest to memory tier.
-      // Assuming if there is no local worker, there are no local blocks in blockInfo.locations.
-      // TODO(cc): Check mContext.hasLocalWorker before finding for a local block when the TODO
-      // for hasLocalWorker is fixed.
-      for (BlockLocation location : blockInfo.getLocations()) {
-        WorkerNetAddress workerNetAddress = location.getWorkerAddress();
-        if (workerNetAddress.getHost().equals(mLocalHostName)) {
-          address = workerNetAddress;
+      List<TieredIdentity> locations = blockInfo.getLocations().stream()
+          .map(location -> location.getWorkerAddress().getTieredIdentity())
+          .collect(Collectors.toList());
+      Optional<TieredIdentity> nearest = mTieredIdentity.nearest(locations);
+      if (nearest.isPresent()) {
+        address = blockInfo.getLocations().stream()
+            .map(BlockLocation::getWorkerAddress)
+            .filter(a -> a.getTieredIdentity() == nearest.get())
+            .findFirst().get();
+        LocalityTier topTier = mTieredIdentity.getTiers().get(0);
+        if (topTier.getTierName().equals(Constants.LOCALITY_NODE)
+            && topTier.getValue().equals(nearest.get().getTiers().get(0).getValue())) {
```

Comment:
Can this logic be abstracted to a method?

Reply:
done